src/exe-elf.c

```diff
@@ -179,7 +187,8 @@ int bxfi_exe_patch_main(bxfi_exe_fn *new_main)
     size_t len = align2_up(offset + sizeof (opcodes), PAGE_SIZE);
 
     mprotect(base, len, PROT_READ | PROT_WRITE | PROT_EXEC);
-    memcpy(nonstd (void *) addr, opcodes, sizeof (opcodes));
+    size_t prelude_len = bxfi_exe_inject_prelude(addr);
```

Comment:
I tried implementing this, but I haven't been able to succeed.

My experience was that I can't really insert nops into the original trampoline, because it interferes with how pseudo-instructions are generated by the assembler. For example, adding a nop just before the `ldr` instruction would generate a different pc-relative jump, which then I would have to patch somehow in the C code when injecting it to a word-aligned Thumb address.

Just for the record, when trying the above, timeout.c began to work (the halfword-aligned one), but all word-aligned tests failed:

```C
if (bxfi_exe_is_arm_func_word_aligned(func_to_patch))
        *trampoline = (void*) (uintptr_t) &bxfi_trampoline_thumb + 2;
```

```
1/4 nested.c          FAIL            0.01s   killed by signal 6 SIGABRT
>>> MALLOC_PERTURB_=72 /tmp/cirrus-ci-build/build/sample/nested.c.bin

2/4 callback.c        FAIL            0.01s   killed by signal 6 SIGABRT
>>> MALLOC_PERTURB_=179 /tmp/cirrus-ci-build/build/sample/callback.c.bin

3/4 context.c         FAIL            0.01s   killed by signal 6 SIGABRT
>>> MALLOC_PERTURB_=195 /tmp/cirrus-ci-build/build/sample/context.c.bin

4/4 timeout.c         OK              2.00s
```

Reply:
I see. It's a bit disappointing that we have to maintain an extra pair of symbols just to delimit a nop, but I can live with it.

Reply:
Alternatively, I wonder if we could hardcode the nop opcode as a constant for ARM and pre-fill the opcodes buffer with a nop slide, and then copying the actual opcodes at an aligned index. I don't prefer one over the other, so your call.

Reply:
We could inject the actual opcode from C, but I thought that would be even more terrible.

Reply:
I'm not sure which way to go, my problem with the hard-coded opcode was portability.
ARM docs state that NOP may be a pseudo-instruction, and the assembler might replace it with different instructions on different processors.

For example, `MOV r8, r8` or `MOV r0, r0`.

Reply:
Yeah, my idea was essentially to do this:

```
    char opcodes[4096];
    bxfi_nop_fill(opcodes, sizeof(opcodes));
    uintptr_t offset = (uintptr_t) base - (uintptr_t) align2((uintptr_t) base, arch_word_size);
    memcpy(opcodes + offset, trampoline, sizeof (opcodes) - offset);
```

and have bxfi_nop_fill fill the buffer with nops depending on the actual architecture, so that we ensure that regardless of the architecture, the trampoline start is word aligned. That said it assumes that the base can't be misaligned by an offset smaller than the size of the smallest nop, but I think the current PR makes the same assumption anyway (with the thumb nop being something like mov r8, r8 and being 2-bytes-long).

Reply:
This sounds good to me.
I'm a bit afraid of the implementation of `bxfi_nop_fill`, so I'd leave this PR as is for now, if you don't mind. :)
When adding the WIndows/iOS support, I'm going to try implementing this as you suggested.

Reply:
Honestly I'm fine with the current code. I'm always happier the better the code is, so if it looks like my suggestion could end up being overall better, of course I'd prefer it, but this is perfectly fine as-is too. And better yet, it works.